tensorflow/core/framework/shape_inference.h

```diff
@@ -889,10 +889,10 @@ inline DimensionOrConstant::DimensionOrConstant(DimensionHandle dim)
 }
 
 inline DimensionOrConstant::DimensionOrConstant(int64_t val) : val(val) {
-  DCHECK(val >= 0 || val == InferenceContext::kUnknownDim)
-      << "Dimension must be non-negative or equal to "
-         "InferenceContext::kUnknownDim but got "
-      << val;
+  if (val < 0 && val != InferenceContext::kUnknownDim) {
+      return errors::InvalidArgument("Dimension must be non-negative or equal to ",
+         "InferenceContext::kUnknownDim but got ", val,".");
+  }
```

Comment:
Actually, you cannot do that here. This is a constructor, see line 194 for the type definition.

What you need to do is to make a static function that can return a value or a status (or return a status after building a outval value) and put the check there. Then callers need to be updated to call that method instead of the constructor.